### ISSUE TITLE
Invalid CSRF token on edit method

### ISSUE THREAD

**Issue by mneil** [FriendsOfCake/crud-view#20]
I originally opened this item in the crud repository but have tracked it down to crud-view. The edit method always results in "Invalid CSRF token" when editing a record.

I did a fresh install of cake 3, added crud, crud-view, bootstrap-ui, and search. Added only the required config. 

Schema _(I've redacted some schema here to make testing easier)_:
```
CREATE TABLE `blog_entries` (
   `id` int(11) NOT NULL AUTO_INCREMENT,
   `featured` tinyint(1) NOT NULL,
   `title` varchar(255) NOT NULL,
   `content` text NOT NULL,
   `created` datetime NOT NULL,
   `updated` timestamp NOT NULL DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP,
   PRIMARY KEY (`id`)
 ) ENGINE=InnoDB AUTO_INCREMENT=5 DEFAULT CHARSET=utf8
```

AppController
```
<?php

namespace App\Controller;

use Cake\Controller\Controller;

class AppController extends Controller
{
  use \Crud\Controller\ControllerTrait;

  public $viewClass = 'CrudView\View\CrudView';

  public $components = [
      'Csrf',
      'Search.Prg',
      'Crud.Crud' => [
        'actions' => [
          'Crud.Index',
          'Crud.Add',
          'Crud.Edit',
          'Crud.View',
          'Crud.Delete'
        ],
        'listeners' => [
          'Crud.RelatedModels',
          'Crud.Redirect',
          'CrudView.View',
          'CrudView.Search',
        ]
      ]
  ];
    public function initialize()
    {
        parent::initialize();
        $this->loadComponent('Flash');
    }
}
```

BlogEntriesController
```
<?php

namespace App\Controller;

use App\Model\Entity\Users;
use Cake\Event\Event;

class BlogEntriesController extends AppController {

  public $paginate = ['limit' => 20];

  public function beforeFilter(Event $event) {
    return parent::beforeFilter($event);
  }

  public function index() {
    return $this->Crud->execute();
  }

  public function add() {
    return $this->Crud->execute();
  }

  public function edit($id) {
    return $this->Crud->execute();
  }
}
```

And finally BlogEntriesTable
```
<?php

namespace App\Model\Table;

use Cake\ORM\Table;
use Cake\Validation\Validator;
use App\Model\Table\Entity;

class BlogEntriesTable extends Table
{

  public function initialize(array $config)
  {
    $this->addBehavior('Search.Search');
    $this->addBehavior('Timestamp', [
        'events' => [
            'Model.beforeSave' => [
                'created' => 'new',
            ]
        ]
    ]);
    parent::initialize($config);
  }

  public function validationDefault(Validator $validator)
  {
    $validator->provider('validation', 'Cake\Validation\Validation');

    return $validator
      ->add('title', [
        'minLength' => [
          'rule' => ['minLength', 10],
          'last' => true,
          'message' => '10 character minimum'
        ],
        'maxLength' => [
          'rule' => ['maxLength', 255],
          'message' => 'Title is too long, 255 character max'
        ]
      ])
      ->add('content', [
          'minLength' => [
              'rule' => ['minLength', 10],
              'last' => true,
              'message' => '10 character minimum'
          ],
          'maxLength' => [
              'rule' => ['maxLength', 5000],
              'message' => 'Content is too long, 5000 character max'
          ]
      ])
      ->add('user_id', 'naturalNumber', [
        'rule' => 'naturalNumber',
        'provider' => 'validation'
      ])
      ->add('blog_category_id', 'naturalNumber', [
        'rule' => 'naturalNumber',
        'provider' => 'validation'
      ]);
  }

}
```

Bootstrap:
```
Plugin::load('Search');
Plugin::load('Crud');
Plugin::load('CrudView');
Plugin::load('BootstrapUI');
```

I can browse to /blog_entries and add a record, no problem. Now edit the record and you'll see "Invalid CSRF token". I've also noticed that the generated token in the form is identical on every request. If anyone sees anything obviously wrong with what I've done above I'd appreciate the help. I'll try and continue to debug this in the meantime. 

Thanks!

**Comment by jadb** [comment]
I don't have time to look at it more right now, but quick (unrelated) tips:

- you don't need to create your actions if you are only `execute()`ing them, 
- validator defaults to the `Cake\Validation\Validation` provider, no need for that line
- timestamp is missing a rule for updating the `updated` field

All that been said, and since you seem to want to save time by using Crud, you should also look into [baking your models, fixtures and related tests](http://book.cakephp.org/3.0/en/bake/usage.html).

If I get time later and you still haven't figured it out or someone else helped, I'll peak some more.

**Comment by mneil** [comment]
Thanks for the response. I'm still digging into this. One thing I've found is that $this->request->data is completely empty when the form is "posted". $_POST contains the data, but $this->request->data is completely empty in both beforeFilter and any action methods.

More insight into your points in case anyone else reads this.

> you don't need to create your actions if you are only execute()ing them,
Thanks, I added this simply because I was working on trying to debug and printing out data in that action helped. It can be removed though in my example above.

>validator defaults to the `Cake\Validation\Validation` provider, no need for that line

Yes and no. It defaults, as far as I can tell, to `Cake\Validation\Validator` and not `Validation`. Validator is missing the `naturalNumber` validation type. Only once I added that provider was I able to use the `naturalNumber`. I don't know if that's important to the issue but it might be worth me trying to remove the provider to test.

> timestamp is missing a rule for updating the updated field
Timestamp is handled in the database. While I know it's convention behind ORM models to control behaviors like this in the codebase I know we won't move away from a SQL model where I won't have the ability to set a current timestamp on UPDATE but thanks again for taking the time to look at it close enough to notice this.

`updated timestamp NOT NULL DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP`

I'm actually moving a Cake 2.4 application over the 3 and am using CRUD since scaffolding has been pulled from the core. I appreciate everyone's work on this library! Hopefully I'l be able to figure out this issue and help some.

Thanks!

**Comment by lorenzo** [comment]
@mneil did you figure out your problem?

**Comment by mneil** [comment]
No I wasn't able to figure out exactly what was happening. I think my original example to recreate the issue is incomplete. I noticed this issue only arises when I'm pulling data from belongsTo relationship which crud-view will automatically create a select element for. If I remove that relationship then it works.

For my original example if you add a blog_categories table and set that to hasMany blog_entries then add a belongsTo relationship to the blog_entries the issue shows up.

I'm only using this behind an admin scaffold and after spending so much time on it I abandoned trying to make it work and just disabled csrf on the admin controller; which all other admin controllers extend. 

I'll try and add a more complete example so it's easier to reproduce but I was able to recreate the issue pretty easily with a fresh install using the example above and then adding that blog_categories relationship.

Thanks!
